### ISSUE TITLE
goto & behaviour doesn't match documentation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8914]
Migrated from [rt.perl.org#43077](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=43077) (status was 'resolved')

Searchable as RT43077$

**Comment by p5pRT** [comment]
### From blgl@hagernas.com

#### Created by blgl@hagernas.com



The documentation for goto & says&#8203;:
 > Instead\, it exits the current subroutine \(losing any changes
 > set by local\(\)\) and immediately calls in its place the named
 > subroutine using the current value of @&#8203;\_\.

Actually\, it reuses the current argument array\, which isn't
necessarily the same as @&#8203;\_\.  The example program below shows
the difference\.

Either goto & should be modified to use @&#8203;\_ or the documentation
should be modified to match reality\.


\#\! /usr/local/bin/perl

use strict;

sub inner
\{
&nbsp;     print "@&#8203;\_\\n";
\}

sub indirect
\{
&nbsp;     \*\_=\[qw\(modified argument list\)\];
&nbsp;     &inner;         \# outputs "modified argument list"
&nbsp;     goto &inner;    \# outputs "original arguments"
\}

indirect\(qw\(original arguments\)\);

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl v5.8.8:

Configured by blgl at Fri Jun  9 19:26:53 CEST 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
   Platform:
     osname=darwin, osvers=8.6.0, archname=darwin-thread- 
multi-64int-2level
     uname='darwin sunwukung.sth.ownit.se 8.6.0 darwin kernel version  
8.6.0: tue mar 7 16:58:48 pst 2006;  
root:xnu-792.6.70.obj~1release_ppc power macintosh powerpc '
     config_args=''
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define  
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=define use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-fno-common -DPERL_DARWIN -no-cpp-precomp - 
fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/ 
include',
     optimize='-Os',
     cppflags='-no-cpp-precomp -fno-common -DPERL_DARWIN -no-cpp- 
precomp -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/ 
usr/local/include'
     ccversion='', gccversion='4.0.1 (Apple Computer, Inc. build  
5247)', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=87654321
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long long', ivsize=8, nvtype='double', nvsize=8,  
Off_t='off_t', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -L/usr/ 
local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-lc
     perllibs=-lc
     libc=/usr/lib/libc.dylib, so=dylib, useshrplib=false,  
libperl=libperl.a
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/ 
usr/local/lib'

Locally applied patches:



@INC for perl v5.8.8:
     /usr/local/lib/perl5/5.8.8/darwin-thread-multi-64int-2level
     /usr/local/lib/perl5/5.8.8
     /usr/local/lib/perl5/site_perl/5.8.8/darwin-thread- 
multi-64int-2level
     /usr/local/lib/perl5/site_perl/5.8.8
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.8.8:
     DYLD_LIBRARY_PATH (unset)
     HOME=/Users/blgl
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/Users/blgl/bin:/Users/blgl/bin/powerpc-apple-darwin:/usr/ 
local/bin:/usr/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/arla/bin
     PERL_BADLANG (unset)
     SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @davidnicol

On 5/30/07\, via RT Bo Lindbergh \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> The documentation for goto & says&#8203;:
>  > Instead\, it exits the current subroutine \(losing any changes
>  > set by local\(\)\) and immediately calls in its place the named
>  > subroutine using the current value of @&#8203;\_\.
>
> Actually\, it reuses the current argument array\, which isn't
> necessarily the same as @&#8203;\_\.  The example program below shows
> the difference\.
>
> Either goto & should be modified to use @&#8203;\_ or the documentation
> should be modified to match reality\.

Here's a demonstration of a way to fake
pushing an alias onto @&#8203;\_\, using pure perl&#8203;:

bash> perl

use Tie&#8203;::Alias;
$X = 12;
sub ListInc\{ \+\+$\_ for @&#8203;\_ \};
sub pushXandgo\{
&nbsp;        my $target = shift;
&nbsp;        tie $\_\[@&#8203;\_\]\, 'Tie&#8203;::Alias' => \\$X;
&nbsp;        goto &$target;
\};
my \($a\,$b\,$c\) = \(9\,10\,11\);
pushXandgo\('ListInc'\,$a\,$b\,$c\);
print "$a $b $c $X\\n";
\_\_END\_\_

I think I'll make some modifications to Tie&#8203;::Alias including better handling
of aliases to tied things and acceptance of

&nbsp;     use Tie&#8203;::Alias $alias => $original;

as invocation syntax\, and re\-release it presently

&nbsp;    use Tie&#8203;::Alias $\_\[@&#8203;\_\] => $NewLastValue

is still a lot clunkier than\, for instance

&nbsp;    push \\\(@&#8203;\_\)\, \\$NewLastValue

but it is pure\, current\, Perl\.

Here's the requested documentation patch&#8203;:


$ diff \-u perl\-5\.9\.4/pod/perlfunc\.pod perl\-5\.9\.4\.modified/pod/perlfunc\.pod
\-\-\- perl\-5\.9\.4/pod/perlfunc\.pod 2006\-08\-15 07&#8203;:37&#8203;:42\.000000000 \-0500
\+\+\+ perl\-5\.9\.4\.modified/pod/perlfunc\.pod        2007\-06\-03
20&#8203;:53&#8203;:24\.000000000 \-0500
@&#8203;@&#8203; \-2357\,10 \+2357\,13 @&#8203;@&#8203;
 doesn't have the stigma associated with other gotos\.  Instead\, it
 exits the current subroutine \(losing any changes set by local\(\)\) and
 immediately calls in its place the named subroutine using the current
\-value of @&#8203;\_\.  This is used by C\<AUTOLOAD> subroutines that wish to
\+argument array\.  This is used by C\<AUTOLOAD> subroutines that wish to
 load another subroutine and then pretend that the other subroutine had
\-been called in the first place \(except that any modifications to C\<@&#8203;\_>
\-in the current subroutine are propagated to the other subroutine\.\)
\+been called in the first place\. Any modifications to the
\+current argument array\, known as C\<@&#8203;\_>\,
\+are propagated to the other subroutine\.  This is also used
\+to implement tail recursion\.
\+
 After the C\<goto>\, not even C\<caller> will be able to tell that this
 routine was called first\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jbenjore

On 5/30/07\, via RT Bo Lindbergh \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Bo Lindbergh
> \# Please include the string&#8203;:  \[perl \#43077\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=43077 >
>
>
> This is a bug report for perl from blgl@&#8203;hagernas\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>
> The documentation for goto & says&#8203;:
>  > Instead\, it exits the current subroutine \(losing any changes
>  > set by local\(\)\) and immediately calls in its place the named
>  > subroutine using the current value of @&#8203;\_\.
>
> Actually\, it reuses the current argument array\, which isn't
> necessarily the same as @&#8203;\_\.  The example program below shows
> the difference\.
>
> Either goto & should be modified to use @&#8203;\_ or the documentation
> should be modified to match reality\.

Better than patching that bit of documentation would be to perhaps
note wherever it is that glob stomping is documented that doing this
to magic variables might remove or mask their magic\.

All the various bits of punctuation variable magic also work the same
way\. If you clobber the glob \(even just temporarilly\) you shouldn't
expect it to still be magic\.

Josh

Josh

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jun 03 20&#8203;:13&#8203;:32 2007\, jjore wrote&#8203;:
> On 5/30/07\, via RT Bo Lindbergh \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > \# New Ticket Created by  Bo Lindbergh
> > \# Please include the string&#8203;:  \[perl \#43077\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=43077 >
> >
> >
> > This is a bug report for perl from blgl@&#8203;hagernas\.com\,
> > generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> >
> > The documentation for goto & says&#8203;:
> >  > Instead\, it exits the current subroutine \(losing any changes
> >  > set by local\(\)\) and immediately calls in its place the named
> >  > subroutine using the current value of @&#8203;\_\.
> >
> > Actually\, it reuses the current argument array\, which isn't
> > necessarily the same as @&#8203;\_\.  The example program below shows
> > the difference\.
> >
> > Either goto & should be modified to use @&#8203;\_ or the documentation
> > should be modified to match reality\.
> 
> Better than patching that bit of documentation would be to perhaps
> note wherever it is that glob stomping is documented that doing this
> to magic variables might remove or mask their magic\.
> 
> All the various bits of punctuation variable magic also work the same
> way\. If you clobber the glob \(even just temporarilly\) you shouldn't
> expect it to still be magic\.

The better solution might be to make goto &foo leave \*\_ alone\, the way
&foo does\.  I believe it would be faster\, too\.  And that would also
allow goto &foo to work inside an eval\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-09\-25T23&#8203;:19&#8203;:38\]
> The better solution might be to make goto &foo leave \*\_ alone\, the way
> &foo does\.  I believe it would be faster\, too\.  And that would also
> allow goto &foo to work inside an eval\.

Seems that way to me\, too\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Sep 26 04&#8203;:47&#8203;:18 2012\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> \[2012\-09\-
> 25T23&#8203;:19&#8203;:38\]
> > The better solution might be to make goto &foo leave \*\_ alone\, the
> way
> > &foo does\.  I believe it would be faster\, too\.  And that would also
> > allow goto &foo to work inside an eval\.
> 
> Seems that way to me\, too\.

Except that we can’t goto &sub from an eval\, as it becomes unclear
whether the sub goneto should trap errors\.

So I have fixed the local \[@&#8203;\*\]\_ problem\, while leaving eval \{ goto &sub
\} as an error\, in commit 049bd5f\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
